# **Data Consistency Checks**

- Find and address mixed type variables in a dataframe
- Find and address missing values in a dataframe
- Find and address duplicate values in a dataframe

In [ ]:
df_ords (Raw Data)
orders_wrangled.csv (Cleaned Data)

In [5]:
# import data
import pandas as pd
import numpy as np 
import os

In [7]:
folderpath = r'C:\Users\Bamidele\Desktop\Training_content\Data Analytics Training - Techbams Solutions\Python\BasketAnalysis_python_project'

In [32]:
df_ords = pd.read_csv( os.path.join(folderpath, '02 Data', 'Cleaned Data', 'orders_wrangled.csv'))

In [38]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [36]:
df_ords.drop(columns = ['Unnamed: 0'], inplace = True)

In [40]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## Mixed Data Type

In [43]:
df_test = pd.DataFrame()

In [49]:
df_test.head()

Empty DataFrame
Columns: []
Index: []

In [51]:
df_test['mix'] = ['a', 'c', 6, True ]

In [53]:
df_test.head()

mix
0     a
1     c
2     6
3  True

In [85]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [61]:
df_test2 = pd.DataFrame()

In [63]:
df_test2['bams'] = [300, 500, 700]

In [65]:
df_test2.head()

bams
0   300
1   500
2   700

In [79]:
for col in df_test2.columns.tolist():
  weird = (df_test2[[col]].map(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test2[weird]) > 0:
    print (col)

In [83]:
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [89]:
df_prods = pd.read_csv( os.path.join(folderpath, '02 Data', 'Raw Data', 'products.csv'), index_col = False)

In [91]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [99]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

- There are 16 missing entries in the product_name column

In [108]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [110]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [164]:
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

## Addressing Missing Values

In [ ]:
Use the describe() function

In [ ]:
# df['columnname].fillna(mean or median value, inplace = True   )

In [ ]:
# df_prods_clean['prices'].fillna(7.1, inplace= True)

In [115]:
df_prods.shape

(49693, 5)

In [117]:
# Create a new prods dataframe that is clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [121]:
df_prods_clean.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [123]:
df_prods_clean.shape

(49677, 5)

 49677 is less than 49693 which is the 16 blanks in product name

In [132]:
df_prods.dropna(subset = ['product_name'], inplace=True)

In [134]:
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [ ]:
# df_bams_clean = df_prods.dropna(subset = ['product_name'])

In [ ]:
df_prods_clean

## Duplicates

In [141]:
df_prods_clean[ df_prods_clean.duplicated()]

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [143]:
df_dups = df_prods_clean[ df_prods_clean.duplicated()]

In [145]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

There are 5 duplicates in the pproducts table

## Addressing Duplicates

In [151]:
df_prods_clean.shape

(49677, 5)

In [153]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [155]:
# Check the shape after removing duplicates
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting the cleaned prods data

In [159]:
df_prods_clean_no_dups.to_csv(os.path.join(folderpath, '02 Data', 'Cleaned Data', 'products_checked.csv'))

# Prepared by Bams